diff --git a/cmake/tests/test_ELSI.f90 b/cmake/tests/test_ELSI.f90
index c6c20b9b4..44b6a0c15 100644
--- a/cmake/tests/test_ELSI.f90
+++ b/cmake/tests/test_ELSI.f90
@@ -1,7 +1,10 @@
       program test
       use elsi
+      use elsi_precision, only: r8
       type(elsi_handle):: eh
       INTEGER,PARAMETER::BLACS_DENSE=0
-      call elsi_init(eh,1,1,BLACS_DENSE,100,50.,50)
+      REAL(r8)::elec
+      elec=50
+      call elsi_init(eh,1,1,BLACS_DENSE,100,elec,50)
       call elsi_finalize(eh)
       end program
